sb/robot/arenas/tin_can_rally_2018.py

```diff
@@ -0,0 +1,244 @@
+
+    @heading.setter
+    def heading(self, new_heading):
+        self._body.angle = new_heading
+
+    def __init__(self, arena, dims, pos=(0.0, 0.0)):
+        self._body = arena._physics_world.create_body(position=(0.0, 0.0),
+                                                      angle=0,
+                                                      type=pypybox2d.body.Body.STATIC)
+        hw = dims[0] / 2.0
+        hh = dims[1] / 2.0
+        fixture = self._body.create_polygon_fixture([
+            (-hw, -hh),
+            (hw, -hh),
+            (hw, hh),
+            (-hw, hh)],
+            restitution=0.2,
+            friction=0.3)
+        self.vertices = fixture.shape.vertices
+        self._w = dims[0]
+        self._h = dims[1]
+        self.location = pos
+        super().__init__(arena)
+
+    def get_corners(self):
+        return [(x + self.location[0], y + self.location[1]) for x, y in self.vertices]
+
+
+class Token(GameObject):
+    grabbable = True
+
+    @property
+    def location(self):
+        return self._body.position
+
+    @location.setter
+    def location(self, new_pos):
+        self._body.position = new_pos
+
+    @property
+    def heading(self):
+        return self._body.angle
+
+    @heading.setter
+    def heading(self, new_heading):
+        self._body.angle = new_heading
+
+    def __init__(self, arena, damping, is_gold=False):
+        self._body = arena._physics_world.create_body(position=(0, 0),
+                                                      angle=0,
+                                                      linear_damping=damping,
+                                                      angular_damping=damping * 2,
+                                                      type=pypybox2d.body.Body.DYNAMIC)
+        super(Token, self).__init__(arena)
+        self.grabbed = False
+        WIDTH = 0.08
+        self._body.create_circle_fixture(
+            radius=WIDTH,
+            restitution=0.2,
+            friction=0.3
+        )
+        self.is_gold = is_gold
+
+    def grab(self):
+        self.grabbed = True
+
+    def release(self):
+        self.grabbed = False
+
+    @property
+    def surface_name(self):
+        return 'sb/{}can{}.png'.format('gold' if self.is_gold else '',
+                                       '_grabbed' if self.grabbed else '')
+
+
+class TCRArena2018(Arena):
+    start_locations = [(-4 + 2.5, -4 + 0.8),
+                       (4 - 2.5, 4 - 0.8)]
+
+    start_headings = [-pi, 0]
+
+    def __init__(self, objects=None):
+        super().__init__(objects)
+        self._init_walls()
+        self._init_tokens()
+
+    def _init_tokens(self):
+        token_locations = [
+            # Numbers from the SVG defined in the rules.
+            (0.5 - 4, 0.8 - 4),
+            (1.2 - 4, 3.7 - 4),
+            (2.0 - 4, 7.7 - 4),
+            (3.4 - 4, 6.0 - 4),
+        ]
+
+        token_locations += [TCRArena2018.rotate(pos, pi) for pos in token_locations]
```

Comment:
Minor: maybe call this as `self.rotate` rather than `TCRArena2018.rotate`? While either works, it's not really important here that it's a class method, though including that makes it more verbose (and had me wondering which class I was currently reading).

Reply:
fixed in 7fa77bb01ce95c89cb13790d64eca5638f2aa90b